base_layer/core/src/transaction.rs

```diff
@@ -84,10 +186,210 @@ pub struct TransactionKernel {
     /// Remainder of the sum of all transaction commitments. If the transaction
     /// is well formed, amounts components should sum to zero and the excess
     /// is hence a valid public key.
-    pub excess: Commitment,
+    pub excess: Option<Commitment<'a>>,
     /// The signature proving the excess is a valid public key, which signs
     /// the transaction fee.
-    pub excess_sig: Signature,
+    pub excess_sig: Option<Signature>,
+}
+
+/// Implementation of the transaction kernel
+impl<'a> TransactionKernel<'a> {
+    /// Creates an empty transaction kernel
+    pub fn empty() -> TransactionKernel<'a> {
+        TransactionKernel { features: KernelFeatures::empty(), fee: 0, lock_height: 0, excess: None, excess_sig: None }
+    }
+
+    /// Build a transaction kernel with the provided fee
+    pub fn with_fee(self, fee: u64) -> TransactionKernel<'a> {
+        TransactionKernel { fee, ..self }
+    }
+
+    /// Build a transaction kernel with the provided lock height
+    pub fn with_lock_height(self, lock_height: u64) -> TransactionKernel<'a> {
+        TransactionKernel { lock_height, ..self }
+    }
+}
+
+/// Implement the canonical hashing function for TransactionKernel for use in ordering
+impl<'a> Hashable for TransactionKernel<'a> {
+    type Hasher = Blake256;
+
+    fn hash(&self) -> Vec<u8> {
+        let mut hasher = Self::Hasher::new();
+        hasher.input(vec![self.features.bits]);
+        let mut fee_lock_height = Vec::new();
+        fee_lock_height.write_u64::<LittleEndian>(self.fee).unwrap();
+        fee_lock_height.write_u64::<LittleEndian>(self.lock_height).unwrap();
+        hasher.input(fee_lock_height);
+        if self.excess.is_some() {
+            hasher.input(self.excess.unwrap().to_bytes());
+        }
+        if self.excess_sig.is_some() {
+            hasher.input(self.excess_sig.unwrap().get_signature().to_bytes());
+        }
+        hasher.result().to_vec()
+    }
+}
+
+hashable_ord!(TransactionKernel);
+
+/// A transaction which consists of a kernel offset and an aggregate body made up of inputs, outputs and kernels.
+pub struct Transaction<'a> {
+    /// This kernel offset will be accumulated when transactions are aggregated to prevent the "subset" problem where
+    /// kernels can be linked to inputs and outputs by testing a series of subsets and see which produce valid
+    /// transactions
+    pub offset: Option<BlindingFactor>,
+    /// The constituents of a transaction which has the same structure as the body of a block.
+    pub body: AggregateBody<'a>,
+    /// reference to the Base point used in the commitments in this transaction
+    pub base: &'a Base,
```

Comment:
Why is the base needed here?

Reply:
This is dictated by the current implementation of the Homomorphic commitment where each commitment takes a reference to a base point pair and the building process of the transaction means that this struct needs to create new commitments so it needs access to the base to do that. I am looking into a better pattern for defining the base.